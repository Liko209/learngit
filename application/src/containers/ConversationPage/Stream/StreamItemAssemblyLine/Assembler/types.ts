/*
 * @Author: Andy Hu
 * @Date: 2019-01-08 10:55:39
 * Copyright Â© RingCentral. All rights reserved.
 */
import { ISortableModel } from '@/store/base/fetch/types';
import { Post } from 'sdk/module/post/entity';
import { StreamItem } from '../../types';
export type AssemblerAddFuncArgs = {
  added: (ISortableModel<Post> & { data: Post })[];
  postList: ISortableModel<Post>[];
  hasMore: boolean;
  streamItemList: _.LoDashImplicitWrapper<StreamItem[]>;
  readThrough: number;
};
export type AssemblerDelFuncArgs = {
  deleted: number[];
  postList: ISortableModel<Post>[];
  streamItemList: _.LoDashImplicitWrapper<StreamItem[]>;
  readThrough: number;
};
export type AssemblerAddFunc = (
  args: AssemblerAddFuncArgs,
) => AssemblerAddFuncArgs;
export type AssemblerDelFunc = (
  args: AssemblerDelFuncArgs,
) => AssemblerDelFuncArgs;
