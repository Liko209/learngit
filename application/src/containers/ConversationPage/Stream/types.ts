/*
 * @Author: Andy Hu
 * @Date: 2018-11-13 18:05:16
 * Copyright Â© RingCentral. All rights reserved.
 */
import { TDelta, ISortableModel } from '../../../store/base/fetch/types';
import PostModel from '../../../store/models/Post';
enum SeparatorType {
  DATE = 'DATE',
  NEW_MSG = 'NEW_MSG',
}

type Separator = {
  type: SeparatorType;
};

type DateSeparator = {
  type: SeparatorType.DATE;
  timestamp: number;
} & Separator;

type NewSeparator = {
  type: SeparatorType.NEW_MSG;
} & Separator;

enum StreamItemType {
  POST = 'POST',
  DATE_SEPARATOR = 'DATE_SEPARATOR',
  NEW_MSG_SEPARATOR = 'NEW_MSG_SEPARATOR',
  INITIAL_POST = 'INITIAL_POST',
}

type BaseElement = {
  type: StreamItemType;
  value: number;
  meta?: any;
};

type StreamItem = {
  id: number;
  type: StreamItemType;
  timeStart: number;
  timeEnd?: number;
  value?: number[];
  meta?: any;
};

type StreamProps = {
  groupId: number;
  viewRef: React.RefObject<any>;
  refresh: () => void;
  jumpToPostId?: number;
};

type StreamViewProps = {
  mostRecentPostId: number;
  firstHistoryUnreadInPage: boolean;
  postIds: number[];
  jumpToPostId: number;
  items: StreamItem[];
  hasMore: (direction: 'up' | 'down') => boolean;
  notEmpty: boolean;
  markAsRead: () => void;
  loadInitialPosts: () => Promise<void>;
  enableNewMessageSeparatorHandler: () => void;
  disableNewMessageSeparatorHandler: () => void;
  handleNewMessageSeparatorState: (event: React.UIEvent<HTMLElement>) => void;
  loadPostUntilFirstUnread: () => Promise<number | undefined>;
  updateHistoryHandler: () => void;
  hasHistoryUnread: boolean;
  clearHistoryUnread: () => void;
  historyUnreadCount: number;
  historyReadThrough?: number;
  firstHistoryUnreadPostId?: number;
  loadInitialPostsError?: Error;
  loading?: boolean;
  lastPost?: PostModel;
  loadMore: (direction: 'up' | 'down', count: number) => Promise<void>;
};

type StreamSnapshot = {
  atBottom: boolean;
  atTop: boolean;
};

type TDeltaWithData = TDelta & {
  added: (ISortableModel & { data: any })[];
};
type ISortableModelWithData = ISortableModel & { data: any };

export {
  StreamSnapshot,
  StreamProps,
  StreamViewProps,
  StreamItemType,
  BaseElement,
  StreamItem,
  SeparatorType,
  Separator,
  DateSeparator,
  NewSeparator,
  TDeltaWithData,
  ISortableModelWithData,
};
