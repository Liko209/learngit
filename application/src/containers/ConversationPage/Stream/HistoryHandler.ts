/*
 * @Author: Valor Lin (valor.lin@ringcentral.com)
 * @Date: 2018-11-14 12:27:09
 * Copyright Â© RingCentral. All rights reserved.
 */
import _ from 'lodash';
import { observable, computed } from 'mobx';
import GroupStateModel from '@/store/models/GroupState';

class HistoryHandler {
  /**
   * The group state when `update()` was called.
   */
  @observable
  groupState: GroupStateModel | null = null;

  /**
   * The latest post when `update()` was called.
   */
  latestPostId: number | null = null;

  /**
   * Remember the current groupState and latestPostId
   */
  update(groupState: GroupStateModel, postIds: number[]) {
    this.groupState = _.cloneDeep(groupState);
    this.latestPostId = _.last(postIds) || null;
  }

  clear() {
    this.groupState = null;
    this.latestPostId = null;
  }

  @computed
  get unreadCount() {
    let count = 0;

    if (this.groupState && this.groupState.unreadCount) {
      count = this.groupState.unreadCount;
    }

    return count;
  }

  @computed
  get readThrough() {
    if (this.groupState) {
      return this.groupState.readThrough;
    }
    return;
  }

  @computed
  get hasUnread() {
    return this.unreadCount > 0;
  }

  getPostsOrderThanLatest(currentPostIds: number[]) {
    const newestPostId = this.latestPostId;
    if (!newestPostId) return [];
    return currentPostIds.filter(id => id <= newestPostId);
  }

  getFirstUnreadPostId(currentPostIds: number[]) {
    if (!this.latestPostId) return;
    const posts = this.getPostsOrderThanLatest(currentPostIds);
    return posts[posts.length - this.unreadCount];
  }

  getDistanceToFirstUnread(currentPostIds: number[]) {
    return (
      this.unreadCount - this.getPostsOrderThanLatest(currentPostIds).length
    );
  }
}

export { HistoryHandler };
