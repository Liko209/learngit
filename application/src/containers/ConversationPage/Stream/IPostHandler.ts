/*
 * @Author: Valor Lin (valor.lin@ringcentral.com)
 * @Date: 2018-10-23 18:25:27
 * Copyright Â© RingCentral. All rights reserved.
 */
import { FetchDataDirection, ISortableModel } from '@/store/base';

interface IPostHandler {
  /**
   * On post added
   */
  onAdded(
    direction: FetchDataDirection,
    addedItems: ISortableModel[],
    allItems: ISortableModel[],
  ): void;
  /**
   * On post deleted
   */
  onDeleted(deletedItems: number[], allItems: ISortableModel[]): void;
}

export { IPostHandler };
