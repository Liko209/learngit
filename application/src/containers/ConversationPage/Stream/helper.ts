/*
 * @Author: Valor Lin (valor.lin@ringcentral.com)
 * @Date: 2018-11-05 14:12:57
 * Copyright Â© RingCentral. All rights reserved.
 */
import { debounce } from 'lodash';
import ReactDOM from 'react-dom';

function getScrollParent(element: HTMLElement, includeHidden: boolean = false) {
  let style = getComputedStyle(element);
  const excludeStaticParent = style.position === 'absolute';
  const overflowRegex = includeHidden
    ? /(auto|scroll|hidden)/
    : /(auto|scroll)/;

  if (style.position === 'fixed') return document.body;
  let parent: HTMLElement | null = element;
  do {
    style = getComputedStyle(parent);
    if (excludeStaticParent && style.position === 'static') {
      continue;
    }

    const overflow = style.overflow || '';
    const overflowX = style.overflowX || '';
    const overflowY = style.overflowY || '';

    if (overflowRegex.test(overflow + overflowX + overflowY)) {
      return parent;
    }
  } while ((parent = parent.parentElement));

  return document.body;
}

async function scrollToComponent(
  component?: React.ReactInstance | null,
  options?: boolean | ScrollIntoViewOptions,
) {
  return new Promise((resolve, reject) => {
    if (!component) return;
    try {
      const el = ReactDOM.findDOMNode(component);
      if (el && el instanceof HTMLElement) {
        el.scrollIntoView(options);
        getScrollParent(el).addEventListener('scroll', debounce(resolve, 100));
      }
    } catch (err) {
      resolve();
      console.warn(err);
    }
  });
}

async function nextTick() {
  return new Promise((resolve: Function) => {
    window.requestAnimationFrame(() => resolve());
  });
}

export { scrollToComponent, nextTick };
