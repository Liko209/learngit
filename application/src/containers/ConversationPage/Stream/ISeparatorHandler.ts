/*
 * @Author: Valor Lin (valor.lin@ringcentral.com)
 * @Date: 2018-10-23 18:25:27
 * Copyright Â© RingCentral. All rights reserved.
 */
import { IPostHandler } from './IPostHandler';
import { Separator } from './types';

interface ISeparatorHandler extends IPostHandler {
  /*
   * When separator is at the same position, the separator that
   * has higher priority would override lower priority one.
   */
  priority: number;
  /*
   * Separators are here
   */
  separatorMap: Map<number, Separator>;
}

export { ISeparatorHandler };
