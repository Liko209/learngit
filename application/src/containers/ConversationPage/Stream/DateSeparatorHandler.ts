/*
 * @Author: Valor Lin (valor.lin@ringcentral.com)
 * @Date: 2018-10-23 18:25:27
 * Copyright Â© RingCentral. All rights reserved.
 */
import _ from 'lodash';
import moment from 'moment';
import { observable, computed } from 'mobx';
import { Post } from 'sdk/models';
import { ISortableModel } from '@/store/base';
import { DateSeparator, SeparatorType } from './types';
import { ISeparatorHandler } from './ISeparatorHandler';
import { QUERY_DIRECTION } from 'sdk/dao';

class DateSeparatorHandler implements ISeparatorHandler {
  priority: number = 1;

  @observable
  private _postsByDateMap: Map<number, ISortableModel<Post>[]> = new Map();

  private _oldestPost?: ISortableModel<Post>;

  @computed
  get separatorMap() {
    const resultMap = new Map<number, DateSeparator>();
    this._postsByDateMap.forEach(
      (posts: ISortableModel<Post>[], timestamp: number) => {
        const firstPost = _.first(posts);
        if (!firstPost) return;
        if (this._oldestPost && firstPost.id === this._oldestPost.id) return;
        resultMap.set(firstPost.id, {
          timestamp,
          type: SeparatorType.DATE,
        });
      },
    );
    return resultMap;
  }

  onAdded(
    direction: QUERY_DIRECTION,
    addedPosts: ISortableModel<Post>[],
    allPosts: ISortableModel<Post>[],
  ): void {
    this._oldestPost = _.first(allPosts);
    _(addedPosts)
      .reverse()
      .forEach((post: ISortableModel<Post>) => {
        if (post.data) {
          const timestamp = moment(post.data.created_at)
            .startOf('day')
            .valueOf();
          const posts = this._postsByDateMap.get(timestamp) || [];
          if (posts) {
            posts.push(post);
          }
          this._postsByDateMap.set(timestamp, _.sortBy(posts, 'id'));
        }
      });
  }

  onDeleted(deletedPostIds: number[], allPosts: ISortableModel<any>[]): void {
    deletedPostIds.forEach((deletedPostId: number) => {
      this._postsByDateMap.forEach(
        (posts: ISortableModel[], timestamp: number) => {
          const newPosts = _.filter(posts, post => post.id !== deletedPostId);
          if (newPosts.length > 0) {
            this._postsByDateMap.set(timestamp, newPosts);
          } else {
            this._postsByDateMap.delete(timestamp);
          }
        },
      );
    });
  }
}

export { DateSeparatorHandler };
