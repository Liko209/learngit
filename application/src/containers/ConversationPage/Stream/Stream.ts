/*
 * @Author: Andy Hu
 * @Date: 2018-09-17 14:01:06
 * Copyright Â© RingCentral. All rights reserved.
 */
import { buildContainer } from '@/base';
import { StreamView } from './Stream.View';
import { StreamViewModel } from './Stream.ViewModel';
import { InfiniteListPlugin } from '@/plugins';
import { StreamProps } from './types';

const Stream = buildContainer<StreamProps>({
  View: StreamView,
  ViewModel: StreamViewModel,
  plugins: [
    new InfiniteListPlugin({
      thresholdUp: 600,
      initialScrollTop: 99999,
      stickTo: 'bottom',
    }),
  ],
});

export { Stream };
