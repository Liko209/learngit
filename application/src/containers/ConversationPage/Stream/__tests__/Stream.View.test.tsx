import React from 'react';
import { shallow } from 'enzyme';
import { ConversationPost } from '../../../ConversationPost';
import GroupStateModel from '@/store/models/GroupState';
import { LoadingMorePlugin } from '@/plugins';
import { StreamViewComponent as StreamView } from '../Stream.View';
import { StreamItemType } from '../types';
import { TimeNodeDivider } from '../../TimeNodeDivider';
import { i18n } from 'i18next';
import { ConversationInitialPost } from '@/containers/ConversationInitialPost';

jest.mock('../../../ConversationSheet', () => ({}));

const baseProps = {
  i18n: {} as i18n,
  tReady: true,
  postIds: [],
  t: () => 'a',
  items: [],
  groupId: 1,
  setRowVisible: jest.fn().mockName('setRowVisible'),
  markAsRead: jest.fn().mockName('markAsRead'),
  atBottom: jest.fn().mockName('atBottom'),
  atTop: jest.fn().mockName('atTop'),
  enableNewMessageSeparatorHandler: jest
    .fn()
    .mockName('enableNewMessageSeparatorHandler'),
  plugins: {
    loadingMorePlugin: new LoadingMorePlugin(),
  },
  hasMoreUp: true,
  hasMoreDown: true,
  notEmpty: true,
  historyGroupState: {} as GroupStateModel,
  historyUnreadCount: 10,
  historyReadThrough: 0,
  hasHistoryUnread: false,
  firstHistoryUnreadInPage: false,
  clearHistoryUnread: jest.fn().mockName('setHasUnread'),
  loadPostUntilFirstUnread: jest.fn().mockName('loadPostUntilFirstUnread'),
  jumpToPostId: 0,
  loadInitialPosts: async () => {},
  updateHistoryHandler: () => {},
  mostRecentPostId: 0,
  resetJumpToPostId: () => null,
  resetAll: (id: number) => {},
};

function renderJumpToFirstUnreadButton({
  hasHistoryUnread,
  firstHistoryUnreadInPage,
  firstHistoryUnreadPostViewed,
}: {
  hasHistoryUnread: boolean;
  firstHistoryUnreadInPage: boolean;
  firstHistoryUnreadPostViewed: boolean;
}) {
  const props = {
    ...baseProps,
    hasHistoryUnread,
    firstHistoryUnreadInPage,
  };

  const wrapper = shallow(<StreamView {...props} />);
  (wrapper.instance() as any)._firstHistoryUnreadPostViewed = firstHistoryUnreadPostViewed;
  wrapper.update();
  const jumpToFirstUnreadButtonWrapper = wrapper.find(
    'JumpToFirstUnreadButtonWrapper',
  );
  const hasJumpToFirstUnreadButton =
    jumpToFirstUnreadButtonWrapper.length === 1;
  return { hasJumpToFirstUnreadButton };
}

describe('StreamView', () => {
  describe('render()', () => {
    it('should render <ConversationPost>', () => {
      const props = {
        ...baseProps,
        postIds: [1, 2],
        items: [
          { type: StreamItemType.POST, value: 1 },
          { type: StreamItemType.POST, value: 2 },
        ],
      };

      const wrapper = shallow(<StreamView {...props} />);
      const card = wrapper.find(ConversationPost);
      const card0 = card.at(0);
      const card1 = card.at(1);
      expect(card).toHaveLength(2);
      expect(card0.props()).toMatchObject({ id: 1, highlight: false });
      expect(card1.props()).toMatchObject({ id: 2, highlight: false });
      expect(card0.key()).toBe('ConversationPost1');
      expect(card1.key()).toBe('ConversationPost2');
    });

    it('should render <TimeNodeDivider>', () => {
      const props = {
        ...baseProps,
        postIds: [1, 2],
        items: [
          { type: StreamItemType.POST, value: 1 },
          { type: StreamItemType.NEW_MSG_SEPARATOR, value: null },
          { type: StreamItemType.POST, value: 2 },
        ],
      };
      const wrapper = shallow(<StreamView {...props} />);
      expect(wrapper.find(ConversationPost)).toHaveLength(2);
      expect(wrapper.find(TimeNodeDivider)).toHaveLength(1);
    });

    describe('hasHistoryUnread=false', () => {
      // JPT-205
      it('should not render jumpToFirstUnreadButton', () => {
        const { hasJumpToFirstUnreadButton } = renderJumpToFirstUnreadButton({
          hasHistoryUnread: false,
          firstHistoryUnreadInPage: false,
          firstHistoryUnreadPostViewed: false,
        });

        expect(hasJumpToFirstUnreadButton).toBeFalsy();
      });
    });

    describe('hasHistoryUnread=true', () => {
      // JPT-206 / JPT-232
      it('should not render jumpToFirstUnreadButton when first history unread in current page and was viewed', () => {
        const { hasJumpToFirstUnreadButton } = renderJumpToFirstUnreadButton({
          hasHistoryUnread: true,
          firstHistoryUnreadInPage: true,
          firstHistoryUnreadPostViewed: true,
        });
        expect(hasJumpToFirstUnreadButton).toBeFalsy();
      });

      // JPT-210
      it('should render jumpToFirstUnreadButton when first history unread in current page but was not viewed', () => {
        const { hasJumpToFirstUnreadButton } = renderJumpToFirstUnreadButton({
          hasHistoryUnread: true,
          firstHistoryUnreadInPage: true,
          firstHistoryUnreadPostViewed: false,
        });

        expect(hasJumpToFirstUnreadButton).toBeTruthy();
      });

      it('should render jumpToFirstUnreadButton when first history unread not in current page', () => {
        const { hasJumpToFirstUnreadButton } = renderJumpToFirstUnreadButton({
          hasHistoryUnread: true,
          firstHistoryUnreadInPage: false,
          firstHistoryUnreadPostViewed: false,
        });
        expect(hasJumpToFirstUnreadButton).toBeTruthy();
      });
    });

    describe('conversationInitialPost', () => {
      function getWrapper(otherProps: object) {
        return shallow(<StreamView {...baseProps} {...otherProps} />);
      }

      it('should render conversationInitialPost when hasMoreUp is false [JPT-478]', () => {
        const hasMoreUp = false;
        const hasSomeMessages = getWrapper({ hasMoreUp, notEmpty: false });
        const noMessages = getWrapper({ hasMoreUp, notEmpty: true });

        expect(hasSomeMessages.find(ConversationInitialPost)).toHaveLength(1);
        expect(noMessages.find(ConversationInitialPost)).toHaveLength(1);
      });

      it('should not render conversationInitialPost when hasMoreUp is true  [JPT-478]', () => {
        const hasMoreUp = true;
        const hasSomeMessages = getWrapper({ hasMoreUp, notEmpty: false });
        const noMessages = getWrapper({ hasMoreUp, notEmpty: true });

        expect(hasSomeMessages.find(ConversationInitialPost)).toHaveLength(0);
        expect(noMessages.find(ConversationInitialPost)).toHaveLength(0);
      });
    });
  });
});
