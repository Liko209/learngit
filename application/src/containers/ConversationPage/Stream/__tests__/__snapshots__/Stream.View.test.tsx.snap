// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamView render() should render container 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StreamView
    items={Array []}
    postIds={Array []}
    setRowVisible={[MockFunction setRowVisible]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": null,
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": null,
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`StreamView render() should render posts and separators 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StreamView
    items={
      Array [
        Object {
          "type": "POST",
          "value": 1,
        },
        Object {
          "type": "NEW_MSG_SEPARATOR",
          "value": null,
        },
        Object {
          "type": "POST",
          "value": 2,
        },
        Object {
          "type": "POST",
          "value": 3,
        },
        Object {
          "type": "POST",
          "value": 4,
        },
      ]
    }
    postIds={
      Array [
        1,
        2,
        3,
        4,
      ]
    }
    setRowVisible={[MockFunction setRowVisible]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Container
          id={1}
        />,
        <NewMessageSeparator />,
        <Container
          id={2}
        />,
        <Container
          id={3}
        />,
        <Container
          id={4}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "id": 1,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "newMessages",
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "id": 2,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "id": 3,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "4",
        "nodeType": "class",
        "props": Object {
          "id": 4,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Container
            id={1}
          />,
          <NewMessageSeparator />,
          <Container
            id={2}
          />,
          <Container
            id={3}
          />,
          <Container
            id={4}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "id": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "newMessages",
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "id": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "id": 3,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "4",
          "nodeType": "class",
          "props": Object {
            "id": 4,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
